### PR TITLE
feat: provide scores in sbom details response

### LINKED ISSUE #952: Dashboard Data - Last 10 SBOMs Barchart
The image below is how V1 looks like in regards of the Dashboard. In the left center part there is BarChart that renders the last 10 SBOMs ingested in the system. Each bar renders the number of vulnerabilities that each SBOM has

image-2024-10-23-11-43-39-294

Here are some things I'd like to have (we can separate each task on its own issue if that works too):

- How to get the last 10 SBOMs  "Ingested"? Currently I can sort SBOMs by `published` but as far as I understand that will sort by the date in  which the SBOM was published which does not necessarily match the date in which it has been ingested into the system.
  - So I can not use `/api/v1/sbom?sort=published:asc` I need something else.
- How to get Vulnerabilities for all 10 SBOMs?
  - Here is the problem I have: If the UI wants to determine vulnerabilities for those 10 SBOMs then I would need to follow these steps:
  - First: Get the last SBOMs with something like `/api/v1/sbom?limit=10&offset=0&sort=published:asc`
  - Second: For each SBOM fetch the advisories available using `/api/v1/sbom/{id}/advisory`. So this endpoint is called 10 times. This endpoint returns a list of advisories associated to each SBOM; each advisory will have a list of Vulnerability IDs (it can be from 0 to hundreds)
  - Third. For each Vulnerability identified in the previous step I need to fetch the vulnerability itself to know its severify and appropriately group them and make sure it is able to be rendered in the BarChart. If we identified hundred of vulnerabilities in the previous step then we will call `/api/v1/vulnerability/{id}` hundred times.

The final math of calls for `/api/v1/vulnerability/{id}` will be = `10SBOM multiplied by #Advisories multiplied by #Vulnerabilities`. This can easily explode from hundreds to thousands. This is not going to scale.

### PR DESCRIPTION
**Pull request by dejanb**
Fixes #952

### CODE REVIEW

**Review by carlosthe19916 — CHANGES_REQUESTED** [review]
@dejanb thanks for putting this PR in place.

Generally speaking this is exactly what I need so I would be happy to have this PR merged but here are some things I've just notice that we might need to address before:

- I've uploaded the files from D3 https://github.com/trustification/trustify/tree/main/etc/datasets/ds3
- Then try to hit the endpoint `http://localhost:8080/api/v1/sbom/{id}/advisory` and the response is huge. The one I got has 40Mb of size which I think is too much.
  - I guess the size of the response could be explained because the response includes `status.vulnerability.advisories` where `advisories` is the field that weights the most. From the top of my mind I would suggest removing `status.vulnerability.advisories` so the response is not that huge but I am not sure if that is fine with you.

**Comment by dejanb** [comment]
Yeah, as mentioned in the issue, I think we should actually return [VulnerabilitySummary](https://github.com/trustification/trustify/blob/main/modules/fundamental/src/vulnerability/model/summary.rs#L17) here, but that needs a bit more changes to fetch for a single vulnerability. It should be faster as well as it will not reach database to fetch unnecessary data.

**Comment by jcrossley3** [comment]
I took a swipe at it, using `VulnerabilityHead` instead of `VulnerabilityDetails` and adding a `severity` field. 

Can you give it a shot @carlosthe19916 and see if we're any closer to what you need? Hoping to add some unit tests tomorrow.

**Comment by carlosthe19916** [comment]
@jcrossley3 Thank you very much!!!!!
This reduced the number of call drastically and I was able to generate the Dashboard using my WIP draft PR https://github.com/trustification/trustify-ui/pull/232

image

- From the pure point of view "It Works". I think YESSS it works your changes gives me back what I need. We can call it DONE. Thanks.
- From the point of long term API:
  - If I hit `/api/v1/sbom/{id}/advisory` then the response is similar to:

```json
[
    {        
        "status": [
            {
                "vulnerability": {
                    "normative": true,
                    "identifier": "CVE-2023-20860",
                    "title": null,
                    "description": null,
                    "reserved": null,
                    "published": null,
                    "modified": null,
                    "withdrawn": null,
                    "discovered": null,
                    "released": null,
                    "cwes": []
                },
                "severity": "high",
                "status": "not_affected",                
            },
```

You can see that the field `severity` is outside of the `vulnerability` object. However, if I hit `/api/v1/vulnerability/{id}` then the severity is in a field named `average_severity` and it is inside the object itself (a vulnerability).

```json
{
  "title": "string", 
  "average_severity": null,
// other fields
}
```

I assume `average` and `average_severity` are the same in this case.

In the long term, the APIs always grow, and having a strategy for having consistency and avoiding fields to be spread in different places and with different names would be necessary for the future. 

For the time being, I am happy to accept this PR as it is because it will allow me to have the Dashboard being implemented! Thank you!

**Comment by jcrossley3** [comment]
> @jcrossley3 Thank you very much!!!!! This reduced the number of call drastically and I was able to generate the Dashboard using my WIP draft PR [trustification/trustify-ui#232](https://github.com/trustification/trustify-ui/pull/232)

Good to hear! :smile: 

> You can see that the field `severity` is outside of the `vulnerability` object. However, if I hit `/api/v1/vulnerability/{id}` then the severity is in a field named `average_severity` and it is inside the object itself (a vulnerability).

I can fix that. I'll flatten all the fields and rename `severity` to `average_severity`.

> For the time being, I am happy to accept this PR as it is because it will allow me to have the Dashboard being implemented! Thank you!

Part of the reason it took me so long to grok this part of the code is its lack of unit tests, so I'd like to fill some in there and clean up the code a bit. Hoping to have that in today, if that's not too long for you to wait for the merge.

**Comment by carlosthe19916** [comment]
I can wait surely. Thanks @jcrossley3 !

**Comment by jcrossley3** [comment]
@carlosthe19916 I refactored things a bit and flattened the structure per your request -- lemme know if it needs further tweaking

@dejanb I would've requested a review from you but since you originated the PR, I can't :shrug:

File: cvss/src/cvss3/score.rs [hunk 1]
```diff
@@ -35,16 +37,12 @@ impl Score {
...
-            Severity::None
-        } else if self.0 < 4.0 {
-            Severity::Low
-        } else if self.0 < 7.0 {
-            Severity::Medium
-        } else if self.0 < 9.0 {
-            Severity::High
-        } else {
-            Severity::Critical
+        match self.0 {
```
Comment (ctron) [inline cvss/src/cvss3/score.rs:40]: :heart:

**Review by ctron — APPROVED** [review]
Love all the improvements!

**Review by dejanb — COMMENTED** [review]
It looks good to me. The UI is currently broken as sbom/vulnerability page still expects the old response. Maybe we should wait with merging until UI catches up? @carlosthe19916

**Comment by carlosthe19916** [comment]
@jcrossley3 Thanks, I've tested it and it works as expected. Whenever you feel it is the time let's merge this PR.

**Review by carlosthe19916 — APPROVED** [review]

**Pull request merged**